runtime/compiler/runtime/JITServerAOTDeserializer.cpp

```diff
@@ -512,74 +571,57 @@ JITServerAOTDeserializer::cacheRecord(const ClassChainSerializationRecord *recor
 
-   uintptr_t offset = (uintptr_t)-1;
-   if (!_sharedCache->isPointerInSharedCache(chain, &offset))
-      {
-      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
-         TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
-            "ERROR: Failed to get SCC offset for class chain %p ID %zu for class %.*s ID %zu",
-            chain, record->id(), ROMCLASS_NAME(ramClasses[0]->romClass), record->list().ids()[0]
-         );
-      return false;
-      }
-   uintptr_t chainLength = chain[0] / sizeof(chain[0]) - 1;
-   uintptr_t *chainItems = chain + 1;
-
-   // Check that the local version of the class chain has the expected length
-   if (record->list().length() != chainLength)
+   // Check that it has the expected length
+   if (ramClassChain.size() != record->list().length())
       {
       if (TR::Options::getVerboseOption(TR_VerboseJITServer))
          TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
             "ERROR: Class chain length mismatch for class %.*s ID %zu: %zu != %zu",
-            ROMCLASS_NAME(ramClasses[0]->romClass), record->list().ids()[0], record->list().length(), chainLength
+            ROMCLASS_NAME(firstClass->romClass), record->list().ids()[0], ramClassChain.size(), record->list().length()
          );
       return false;
       }
 
-   // Validate each ROMClass in the chain
-   for (size_t i = 0; i < chainLength; ++i)
+   // Validate each class in the server's chain (which should all be cached by now)
+   for (size_t i = 0; i < ramClassChain.size(); ++i)
       {
-      J9ROMClass *romClass = _sharedCache->romClassFromOffsetInSharedCache(chainItems[i]);
-      if (romClass != ramClasses[i]->romClass)
+      auto ramClass = getRAMClass(record->list().ids()[i], comp, wasReset);
+      if (!ramClass)
+         return false;
+
+      if (ramClass != ramClassChain[i])
          {
          if (TR::Options::getVerboseOption(TR_VerboseJITServer))
             TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
                "ERROR: Class %.*s mismatch in class chain ID %zu for class %.*s ID %zu",
-               ROMCLASS_NAME(ramClasses[i]->romClass), record->id(),
-               ROMCLASS_NAME(ramClasses[0]->romClass), record->list().ids()[0]
+               ROMCLASS_NAME(ramClassChain[i]->romClass), record->id(),
+               ROMCLASS_NAME(firstClass->romClass), record->list().ids()[0]
             );
          return false;
          }
       }
 
-   _classChainMap.insert(it, { record->id(), offset });
+   // Chain validation is now complete
+   size_t chainSize = (record->list().length() + 1) * sizeof(uintptr_t); // offset chains are expected to store their size in bytes in their first entry
+   auto deserializerChain = (uintptr_t *)TR::Compiler->persistentGlobalMemory()->allocatePersistentMemory(chainSize);
+   if (!deserializerChain)
+      throw std::bad_alloc();
+
+   deserializerChain[0] = chainSize;
+   memcpy(deserializerChain + 1, record->list().ids(), record->list().length());
```

Comment:
This is incorrect, incidentally. The `idAndType` should be written to the chain, not the IDs themselves.